src/epoch.rs

```diff
@@ -0,0 +1,82 @@
+/// The global epoch
+///
+/// The last bit in this number is unused and is always zero. Every so often the global epoch is
+/// incremented, i.e. we say it "advances". A pinned mutator may advance the global epoch only if
+/// all currently pinned mutators have been pinned in the current epoch.
+///
+/// If an object became garbage in some epoch, then we can be sure that after two advancements no
+/// mutator will hold a reference to it. That is the crux of safe memory reclamation.
+
+use std::ops::Deref;
+use std::sync::atomic::AtomicUsize;
+use std::sync::atomic::Ordering::{Relaxed, Acquire, Release, SeqCst};
+
+use mutator::Registry;
+use mutator::Scope;
+use sync::list::{List, IterResult};
+use util::cache_padded::CachePadded;
+
+/// The global epoch is a (cache-padded) integer.
+#[derive(Default, Debug)]
+pub struct Epoch {
+    epoch: CachePadded<AtomicUsize>,
+}
+
+impl Epoch {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Attempts to advance the global epoch.
+    ///
+    /// The global epoch can advance only if all currently pinned mutators have been pinned in the
+    /// current epoch.
+    ///
+    /// Returns the current global epoch.
+    #[cold]
+    pub fn try_advance<'scope>(&self, registries: &List<Registry>, scope: &Scope) -> usize {
+        let epoch = self.epoch.load(Relaxed);
+        ::std::sync::atomic::fence(SeqCst);
```

Comment:
@stjepang @Vtec234 I think SC reads/writes are not sufficient to achieve the required synchronization.  For example, Crossbeam needs such synchronization:

```
clode block X;
read global epoch G
||
read global epoch G+1
code block Y;     // should know of the code block X, as if release/acquire synchronization is done
```

According to a recent understanding of SC semantics (http://plv.mpi-sws.org/scfix/), SC reads/writes cannot provide such guarantee, and we should use SC fence's "cumulativity" for that purposes.

I am clarifying this point in the upcoming RFC (https://github.com/jeehoonkang/crossbeam-rfcs/blob/relaxed-memory/text/2017-07-23-relaxed-memory.md).  Shall we discuss the orderings and correctness in that RFC?